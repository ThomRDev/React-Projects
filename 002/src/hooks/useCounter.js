import { useState } from "react"
const useCounter = (initialState) =>{
    const [counter, setCounter] = useState(initialState);
    const increment = (factor = 1) => setCounter(counter + factor)
    const decrement = (factor = 1) => {
        if(counter >= 1) setCounter(counter - factor) 
    }
    const reset = () => setCounter(initialState)


    return {
        counter,
        actions : {
            increment,
            decrement,
            reset
        }
    }

}

export default useCounter