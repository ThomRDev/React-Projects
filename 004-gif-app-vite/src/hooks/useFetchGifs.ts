import { useEffect, useState } from "react"
import { getGifs, IGif} from "../helpers/getGif"

export const useFetchGifs = (value : string)=>{
    const [state,setState] = useState({ data : [] as IGif[], loading : true})
    useEffect(()=>{
        getGifs(value)
        .then(data=>{
            // data?.length &&
            setTimeout(()=>{
                setState({
                    data,
                    loading:false
                })
            },1000)
        })
    },[value])
    return state
}

