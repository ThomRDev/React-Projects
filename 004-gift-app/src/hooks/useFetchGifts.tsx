import { useEffect, useState } from "react"
import { getGif } from "../helpers/getGif"
// export interface DataHook{
//     data : any[],
//     loading : boolean
// }

export const useFetchGifts = (value : string)=>{
    const [state,setState] = useState({ data : [] as object[], loading : true})
    useEffect(()=>{
        getGif(value)
        .then(data=>{
            // data?.length &&
            setTimeout(()=>{
                setState({
                    data,
                    loading:false
                })
            },1000)
        })
        // eslint-disable-next-line
    },[value])
    return state
}

