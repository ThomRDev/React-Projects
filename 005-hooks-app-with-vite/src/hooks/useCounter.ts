import { useState } from "react"

export const useCounter = (initialValue = 0) => {
  const [state,setState] = useState(initialValue)
  const increment = (factor = 1) => setState(state + factor) 
  const subtract = (factor = 1) => setState(state - factor) 
  const reset = () => setState(initialValue)
  return {
    state,
    increment,
    subtract,
    reset
  } 
}