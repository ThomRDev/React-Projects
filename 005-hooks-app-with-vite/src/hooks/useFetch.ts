import { useEffect, useRef, useState } from "react"

const getData = async (url:string) => {
  const response = await fetch(url)
  const data = await response.json()
  return data 
}


export const useFetch = (url:string) => {
  const ref = useRef(true)

  const [state,setState] = useState({
    // cambiar esto de any
    data : [] as any,
    isLoading : true,
    hasError : false
  })

  useEffect(()=>{
    ref.current = true
    setState({
      data : [],
      isLoading: true,
      hasError : false
    })
    getData(url)
    .then((data) => {
      if(ref.current){
        setState({
          data,
          isLoading: false,
          hasError : false
        })
      }
    }).catch((err) => {
      if(ref.current){
        setState({
          data : [],
          isLoading: false,
          hasError : true
        })
      }
    });

    return () => {
      ref.current = false
    }
  },[url])

  return state
}
