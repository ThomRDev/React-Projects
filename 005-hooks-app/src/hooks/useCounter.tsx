import { useState } from "react"

export const useCounter = (initialState = 0) =>{
    const [counter,setCounter] = useState(initialState)
    
    const decrement = (factor = 1)=>{
        setCounter(counter - factor)
    }
    const increment = (factor = 1)=>{
        setCounter(counter + factor)
    }
    const reset = ()=>setCounter(initialState)
    
    return {
        counter,
        actions:{
            "decrement" : decrement,
            "increment" : increment,
            "reset" : reset,
        }
    }
}