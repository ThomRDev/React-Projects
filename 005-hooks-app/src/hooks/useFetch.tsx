import { useEffect, useRef, useState } from "react"


const timer = (delay:number) => new Promise<void>((resolve)=>{
    setTimeout(()=>{
        resolve()
    },delay)
})

async function getData(url:string){
    const result = await fetch(url)
    await timer(2000)
    const data = await result.json()
    return { ...data } 
}

export const useFetch = (url:string) => {
    
    // la idea de esto es mantener una referencia cuando el hook esta vivo o el componente que usa el hook esta montado
    const isMounted = useRef(true)
    const [state,setState] = useState({
        error : null,
        loading : true,
        data : null
    } as { error:null | boolean,data : null | any,loading : boolean})
    
    useEffect(()=>{
        return () => {
            isMounted.current  = false
        }
    },[])
    
    useEffect(()=>{
        isMounted.current = true
        setState({
            data : null,
            loading : true,
            error : false
        })
        getData(url)
        .then(data=>{
            console.log("async");
            if(isMounted.current){
                console.log("async ok");
                setState({
                    data,
                    loading : false,
                    error : false
                })
            }
        })
        .catch(err=>{
            if(isMounted.current){
                setState({
                    data:null,
                    error : true,
                    loading : false
                })
            }
        })

        return () => {
            console.log("desmontando [url]");
            isMounted.current = false
        }
    // eslint-disable-next-line
    },[url])

    return state
}

