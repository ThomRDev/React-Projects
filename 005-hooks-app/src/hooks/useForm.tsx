import { useState } from "react"

const useForm = ({...initFormValues}) => {
    const [formValues,setFormValues] = useState(initFormValues)
    
    const reset = () => setFormValues(initFormValues) 
    
    const handleChangeInput = (event : any) =>{
        event?.preventDefault()
        const { name,value} = event?.target
        setFormValues({
            ...formValues,
            [name] : value
        })
    }

    return { ...formValues,reset,handleChangeInput }
}

export default useForm