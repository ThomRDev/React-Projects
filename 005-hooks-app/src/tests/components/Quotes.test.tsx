import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer"
import QuotesApp from "../../components/02-useEffect/Quotes"
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";

jest.mock("../../hooks/useFetch")
jest.mock("../../hooks/useCounter")

// esto es solo valido en typescript
// en js solo se usa el hook con su nombre
const mockUseFetch = useFetch as jest.MockedFunction<typeof useFetch>;
const mockUseCounter = useCounter as jest.MockedFunction<typeof useCounter>;

describe('Test in Quotes.tsx', () => {

  beforeEach(()=>{
    mockUseCounter.mockReturnValue({
      counter : 10,
      actions:{
        increment : () => {},
        decrement : () => {},
        reset : () => {},
      }
    })
  })

  test('spapshot', () => {
    mockUseFetch.mockReturnValue({
        data  :null,
        loading : true,
        error : null
    })
    const tree = renderer.create(<QuotesApp />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('must show the information', () => {
    mockUseFetch.mockReturnValue({
      data  : [{
        author :"Author",
        quote : "Quote - Text"
      }],
      loading : false,
      error : null
    })
    render(<QuotesApp />)
    
    expect(screen.getByText("Author")).toBeInTheDocument()
    // screen.debug()
    // expect(screen.getByRole('heading')).toHaveTextContent('Welcome, John Doe')
    
    // sin enzime nose
    // const wrapper = shallow(<QuotesApp />)
    // expect(wrapper.find(".alert").exists()).toBe(false)
    // expect(wrapper.find(".mb-0").text().trim()).toBe("Author")
    // expect(wrapper.find("footer").text().trim()).toBe("Quote - Text")
    
  })
})