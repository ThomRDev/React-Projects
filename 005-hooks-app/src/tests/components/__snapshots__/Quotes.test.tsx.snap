// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in Quotes.tsx spapshot 1`] = `
<div>
  <div
    className="alert alert-info text-center"
  >
    loading ... 
  </div>
  <button
    className="btn btn-primary disabled"
    disabled={true}
    onClick={[Function]}
  >
    Cambiar
  </button>
</div>
`;
